src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -2915,6 +2904,40 @@ void G1CollectedHeap::gc_tracer_report_gc_end(bool concurrent_operation_is_full_
   _gc_timer_stw->time_partitions());
 }
 
+// GCTraceTime wrapper that constructs the message according to GC pause type and
+// GC cause.
+class G1YoungGCTraceTime {
+  G1GCPauseType _pause_type;
+
+  static const uint MaxYoungGCNameLength = 128;
+  char _young_gc_name_data[MaxYoungGCNameLength];
+
+  GCTraceTime(Info, gc) _tt;
+
+  char* update_young_gc_name() {
+    snprintf(_young_gc_name_data,
+             MaxYoungGCNameLength,
+             "Pause Young (%s)%s",
+             G1GCPauseTypeHelper::to_string(_pause_type),
+             G1CollectedHeap::heap()->evacuation_failed() ? " (Evacuation Failure)" : "");
+    return _young_gc_name_data;
+  }
+
+public:
+
+  G1YoungGCTraceTime(GCCause::Cause cause) :
+    // Take snapshot of current pause type at start as it may be modified during gc.
+    // The strings for all Concurrent Start pauses are the same, so the parameter
+    // does not matter here.
+    _pause_type(G1CollectedHeap::heap()->collector_state()->young_gc_pause_type(false /* concurrent_operation_is_full_mark */)),
+    _tt(update_young_gc_name(), NULL, cause, true) {
+  }
+
+  ~G1YoungGCTraceTime() {
+    update_young_gc_name();
+  }
```

Comment:
Do I understand correctly that we update the name in _young_gc_name_data in the destructor, and then depend on that _tt does not  copy that value earlier --- only the pointer? This seems quite fragile if GCTraceTimeWrapper would change, and I think would at least warrant a comment in the constructor that the string *will* be updated in the destructor.

Reply:
Added. Note that the previous code also had that assumption, this is why the `young_gc_name` array needed to be alive the whole method.
The added test will detect if it does not work any more.